File: backend/app/api/routes/audit.py
import uuid
from typing import Any

from fastapi import APIRouter, Depends, HTTPException

from app.api.deps import SessionDep, require_admin
from app.models import AuditLogPublic, AuditLogsPublic
from app.services.audit import AuditService

router = APIRouter()


@router.get("/", response_model=AuditLogsPublic, dependencies=[Depends(require_admin)])
def read_audit_logs(
    session: SessionDep,
    skip: int = 0,
    limit: int = 100,
    user_name: str | None = None,
    action: str | None = None,
    entity_type: str | None = None,
    search: str | None = None,
) -> Any:
    """
    Retrieve audit logs. Only admin can access.
    """
    service = AuditService(session)

    # Get audit logs with user info efficiently loaded
    audit_logs_public, count = service.get_audit_logs_with_users(
        skip=skip,
        limit=limit,
        user_name=user_name,
        action=action,
        entity_type=entity_type,
        search=search,
    )

    return AuditLogsPublic(data=audit_logs_public, count=count)


@router.get("/{audit_log_id}", response_model=AuditLogPublic, dependencies=[Depends(require_admin)])
def read_audit_log(
    session: SessionDep,
    audit_log_id: uuid.UUID,
) -> Any:
    """
    Get audit log by ID. Only admin can access.
    """
    service = AuditService(session)

    audit_log = service.get_audit_log_with_user(audit_log_id)
    if not audit_log:
        raise HTTPException(status_code=404, detail="Audit log not found")

    return audit_log


@router.get("/stats/summary", dependencies=[Depends(require_admin)])
def get_audit_stats(
    session: SessionDep,
) -> Any:
    """
    Get audit statistics summary. Only admin can access.
    """
    service = AuditService(session)
    return service.get_audit_stats(days=30)

File: frontend/src/api/audit.ts
import type { AuditLogPublic } from "@/client/types.gen"
import { apiClient } from "@/lib/axios"

interface PaginatedResponse<T> {
  data: T[]
  count: number
}

interface AuditStatsResponse {
  period_days: number
  total_actions: number
  actions_by_type: Record<string, number>
  actions_by_entity: Record<string, number>
  top_users: Record<string, number>
}

/**
 * Get list of audit logs with filtering and pagination
 */
export async function getAuditLogs(params?: {
  skip?: number
  limit?: number
  user_name?: string | null
  action?: string | null
  entity_type?: string | null
  search?: string | null
}): Promise<PaginatedResponse<AuditLogPublic>> {
  const response = await apiClient.get<PaginatedResponse<AuditLogPublic>>(
    "/api/v1/audit/",
    {
      params: {
        skip: params?.skip || 0,
        limit: params?.limit || 50,
        user_name: params?.user_name || undefined,
        action: params?.action || undefined,
        entity_type: params?.entity_type || undefined,
        search: params?.search || undefined,
      },
    },
  )
  return response.data
}

/**
 * Get single audit log by ID
 */
export async function getAuditLog(auditLogId: string): Promise<AuditLogPublic> {
  const response = await apiClient.get<AuditLogPublic>(
    `/api/v1/audit/${auditLogId}`,
  )
  return response.data
}

/**
 * Get audit statistics summary
 */
export async function getAuditStats(): Promise<AuditStatsResponse> {
  const response = await apiClient.get<AuditStatsResponse>(
    "/api/v1/audit/stats/summary",
  )
  return response.data
}
